{"cells":[{"cell_type":"markdown","id":"204117e6","metadata":{},"source":"## The following code shows the analyses of genomap and genoNet for five different tasks on four sets of scRNA-seq data"},{"cell_type":"code","source":"# For the tasks below, four datasets analysed in the manuscript will be automatically loaded. \r\n# However, you can upload your own dataset, create genomaps and train the supervised or unsupervised\r\n# genoNet models for different tasks. \r\n# Our data were saved as .mat file to reduce the data size (normally .csv file needs more disk space). \r\n# However, .csv files can also be loaded in the way shown in the third section","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"id":"d2a84f36","metadata":{},"outputs":[],"source":"# Load all necessary python packages needed for the reported analyses\n# in our manuscript\n\n%matplotlib inline\n\nimport pandas as pd\nimport matplotlib.pyplot as plt\nimport numpy as np\nimport scipy\nimport scipy.io as sio\nimport numpy as np\nimport scipy.io as sio\nimport genoNet as gNet\nimport os\nimport torch\nfrom torch.utils.data import DataLoader, Dataset\nfrom pytorch_grad_cam import GradCAM\nfrom pytorch_grad_cam.utils.model_targets import ClassifierOutputTarget\nfrom tqdm import tqdm\nimport sklearn\nimport phate\nfrom sklearn.manifold import TSNE\nfrom pyclustering.cluster.xmeans import xmeans\nfrom pyclustering.cluster.center_initializer import kmeans_plusplus_initializer\nfrom pyclustering.cluster import cluster_visualizer\n\nfrom genoNet import geneDataset,_get_device,load_genoNet,predict,traingenoNet,rescale\nfrom genomap import construct_genomap\n\n# For reproducibility\ntorch.manual_seed(1)\ntorch.use_deterministic_algorithms(True)"},{"cell_type":"markdown","id":"aa69c03c","metadata":{},"source":"## Create genomaps from Tabula Muris (TM) scRNA-seq dataset"},{"cell_type":"code","execution_count":null,"id":"f1316f83","metadata":{},"outputs":[],"source":"# First, we load the TM data. Data should be in cells X genes format, \n# i.e., each row should correspond to gene expressions of a cell and each column\n# should represent the expression value of a specific gene. \n# The original TM data is very large with 54,865 cells and 19,791 genes. \n# We select the 1089 most variable genes from the data to reduce its size. \n# The reason behind chosing 1089 is that it is a square number (33*33)\n# However, you can choose any other number. The following commented 4-lines code \n# allows one to select the most variable genes (numGene=1089 was used in our anlaysis)\n# varX=np.var(dataFull,axis=0)\n# idxV=np.argsort(varX)\n# idxVX=np.flip(idxV)\n# dataReduced=dataFull[:,idxVX[0:numGene]]\n\ndata = pd.read_csv('../data/TMdata.csv', header=None,\n                   delim_whitespace=False)\n\n# Creation of genomaps\n# Selection of row and column number of the genomaps \n# To create square genomaps, the row and column numbers are set to be the same.\ncolNum=33 \nrowNum=33\n# Normalization of the data\ndataNorm=scipy.stats.zscore(data,axis=0,ddof=1)\n# Construction of genomaps\ngenoMaps=construct_genomap(dataNorm,rowNum,colNum,epsilon=0.0,num_iter=200)\n\n# Visualization of the constructed genomaps:\n# The \"genoMaps\" array: All the constructed genomaps are saved in the array. This \n# array has four indices. The first one indexes the series of genomaps. \n# The second and third ones denote the row and column number in a genomap.\n# The fourth index is introduced to facillitate the caclulation using Pytorch or Tensorflow  \n# to visualize  the i-th genomap, set the first index variable to i (i=10 here) \nfindI=genoMaps[10,:,:,:]\nplt.figure(1)\nplt.imshow(findI, origin = 'lower',  extent = [0, 10, 0, 10], aspect = 1)\nplt.title('Genomap of a cell from TM dataset')"},{"cell_type":"markdown","id":"d68dc5bd","metadata":{},"source":"## Train and test the genoNet on the genomaps of TM dataset for cell recognition"},{"cell_type":"code","execution_count":null,"id":"5b86323b","metadata":{},"outputs":[],"source":"# Load ground truth cell labels of the TM dataset\ngt_data = sio.loadmat('../data/GT_TM.mat')\nGT = np.squeeze(gt_data['GT'])\n\n# Load the indices of training and test data (70% training, 30% testing data)\nindex_data = sio.loadmat('../data/index_TM.mat')\nindxTest = np.squeeze(index_data['indxTest'])\nindxTrain = np.squeeze(index_data['indxTrain'])\n\nGT = GT - 1 # to ensure the labels begin with 0 to conform with PyTorch\n\n# Split the data for training and testing\ndataMat_CNNtrain = genoMaps[indxTrain-1,:,:,:] # to ensure the index starts at 0 to conform with\n# python indexing\ndataMat_CNNtest = genoMaps[indxTest-1,:,:,:]\ngroundTruthTest = GT[indxTest-1]\ngroundTruthTrain = GT[indxTrain-1]\nclassNum = len(np.unique(groundTruthTrain))\n\n# Preparation of training and testing data for PyTorch computation\nXTrain = dataMat_CNNtrain.transpose([0,3,1,2])\nXTest = dataMat_CNNtest.transpose([0,3,1,2])\nyTrain = groundTruthTrain\nyTest = groundTruthTest\n\n\n# Train the network in PyTorch framework\nminiBatchSize = 128\nnet = gNet.traingenoNet(XTrain, yTrain, maxEPOCH=150, batchSize=miniBatchSize, verbose=True)\ntorch.save(net.state_dict(),'../results/genoNet_TM.pt')\n\n# Process the test data\ntestset = gNet.geneDataset(XTest, yTest)\ntestloader = gNet.DataLoader(testset, batch_size=miniBatchSize, shuffle=False)\ndevice = gNet._get_device()\n\n# Perform cell classification/reocognition\nprob_test = gNet.predict(net, testloader, device)\npred_test = np.argmax(prob_test, axis=-1)\n\n# Compute the accuracy of cell classification/reocognition\nprint('Classification accuracy of genomap+genoNet for TM dataset:'+str(np.sum(pred_test==yTest) / pred_test.shape[0]))"},{"cell_type":"markdown","id":"e33601f5","metadata":{},"source":"## Discovery of the cell- and class-specific gene sets from TM dataset"},{"cell_type":"code","execution_count":null,"id":"d6cbd6be","metadata":{},"outputs":[],"source":"# The trained genoNet in the last section is used here to find the cell- and class-specific \n# gene sets. We use the GRAD-CAM technique to compute the class activation maps \n# of the trained genoNet for given genomaps.\n# The class activation map shows which pixels in the genomaps are important\n# for decision-making of the genoNet.\n# We emphasize that each genomap pixel represents \n# a gene and the class activation value of the pixel\n# denotes the importance of that particular gene in cell classification\ntarget_layers = [net.conv1]\n# Grad cam object\ncam = GradCAM(model=net, target_layers=target_layers, use_cuda=False)\n\n# We compute the class activation maps of all the cells\n# of a specific cell type to find the class-specific important genes.\n# Load the indices of different cell types\ngt_data = sio.loadmat('../data/index_GRADCAM.mat')\n\n# Compute class activation maps of B-cells using GRAD-CAM\nindxcell=gt_data['idxBcell']\n# Create a numpy array to store the grad CAM results\ngrayscale_cam=np.zeros([indxcell.shape[0],colNum*rowNum])\n\nfor i in range(indxcell.shape[0]):\n    # Looping over all B-cells to compute the class activation maps\n    grad_cam_img = torch.Tensor(dataMat_CNNtrain[i,:,:,:].transpose(2,0,1)\n                            .reshape((1,1,colNum,rowNum))).to(device)\n    pred_label = torch.argmax(net(grad_cam_img),axis=-1)\n    targets = [ClassifierOutputTarget(pred_label)]\n    A = cam(input_tensor=grad_cam_img, targets=targets)\n    B=np.reshape(A.T,(1,colNum*rowNum))\n    grayscale_cam[i,:]=B\n\n# Plot an example activation map of a B-cell, which highlights the genes that are important \n# for the decision made by the genoNet.\nplt.figure(2)\nplt.imshow(np.squeeze(A))\nplt.title('Class activation map of a B-cell')\n\n# Compute the mean of gene activation values across all the B-cells\n# to find the average activity of all the genes in B-cells\ngraySmeanB=np.mean(grayscale_cam,axis=0) \n\n\n# Repeat the same procedure for T cell, natural killer cells,\n# Monocytes, Pneumoctyes and Hematopoitic cells\n\nindxcell=gt_data['idxTcell']\ngrayscale_cam=np.zeros([indxcell.shape[0],colNum*rowNum])\nfor i in range(indxcell.shape[0]):\n\n    grad_cam_img = torch.Tensor(dataMat_CNNtrain[i,:,:,:].transpose(2,0,1)\n                            .reshape((1,1,colNum,rowNum))).to(device)\n    pred_label = torch.argmax(net(grad_cam_img),axis=-1)\n    targets = [ClassifierOutputTarget(pred_label)]\n    A = cam(input_tensor=grad_cam_img, targets=targets)\n    B=np.reshape(A.T,(1,colNum*rowNum))\n    grayscale_cam[i,:]=B\n\ngraySmeanT=np.mean(grayscale_cam,axis=0) # Activity of genes in T-cells\n\nindxcell=gt_data['idxNK']\ngrayscale_cam=np.zeros([indxcell.shape[0],colNum*rowNum])\nfor i in range(indxcell.shape[0]):\n\n    grad_cam_img = torch.Tensor(dataMat_CNNtrain[i,:,:,:].transpose(2,0,1)\n                            .reshape((1,1,colNum,rowNum))).to(device)\n    pred_label = torch.argmax(net(grad_cam_img),axis=-1)\n    targets = [ClassifierOutputTarget(pred_label)]\n    A = cam(input_tensor=grad_cam_img, targets=targets)\n    B=np.reshape(A.T,(1,colNum*rowNum))\n    grayscale_cam[i,:]=B\n\ngraySmeanNK=np.mean(grayscale_cam,axis=0) # Activity of genes in natural killer cells\n\nindxcell=gt_data['idxMono']\ngrayscale_cam=np.zeros([indxcell.shape[0],colNum*rowNum])\nfor i in range(indxcell.shape[0]):\n\n    grad_cam_img = torch.Tensor(dataMat_CNNtrain[i,:,:,:].transpose(2,0,1)\n                            .reshape((1,1,colNum,rowNum))).to(device)\n    pred_label = torch.argmax(net(grad_cam_img),axis=-1)\n    targets = [ClassifierOutputTarget(pred_label)]\n    A = cam(input_tensor=grad_cam_img, targets=targets)\n    B=np.reshape(A.T,(1,colNum*rowNum))\n    grayscale_cam[i,:]=B\n\ngraySmeanMono=np.mean(grayscale_cam,axis=0) # Activity of genes in Monocytes\n\n\nindxcell=gt_data['idxPne']\ngrayscale_cam=np.zeros([indxcell.shape[0],colNum*rowNum])\nfor i in range(indxcell.shape[0]):\n\n    grad_cam_img = torch.Tensor(dataMat_CNNtrain[i,:,:,:].transpose(2,0,1)\n                            .reshape((1,1,colNum,rowNum))).to(device)\n    pred_label = torch.argmax(net(grad_cam_img),axis=-1)\n    targets = [ClassifierOutputTarget(pred_label)]\n    A = cam(input_tensor=grad_cam_img, targets=targets)\n    B=np.reshape(A.T,(1,colNum*rowNum))\n    grayscale_cam[i,:]=B\ngrayscale_cam=np.nan_to_num(grayscale_cam)\ngraySmeanPne=np.mean(grayscale_cam,axis=0) # Activity of genes in Pneumoctyes\n\n\nindxcell=gt_data['idxHema']\ngrayscale_cam=np.zeros([indxcell.shape[0],colNum*rowNum])\nfor i in range(indxcell.shape[0]):\n\n    grad_cam_img = torch.Tensor(dataMat_CNNtrain[i,:,:,:].transpose(2,0,1)\n                            .reshape((1,1,colNum,rowNum))).to(device)\n    pred_label = torch.argmax(net(grad_cam_img),axis=-1)\n    targets = [ClassifierOutputTarget(pred_label)]\n    A = cam(input_tensor=grad_cam_img, targets=targets)\n    B=np.reshape(A.T,(1,colNum*rowNum))\n    grayscale_cam[i,:]=B\n\ngraySmeanHema=np.mean(grayscale_cam,axis=0) # Activity of genes in Hematopoitic cells\n\n\n# Stack all the average gene activity values in different types of cells for plotting\nactivityMatrix=np.vstack((graySmeanB,graySmeanHema,graySmeanMono,graySmeanNK,graySmeanT,graySmeanPne))\n\n# Load the gene names for plotting\ngeneNames = pd.read_csv('../data/gene_interest_ROI_ImXreXB.csv', header=None,\n                   delim_whitespace=False)\n\n\n# The following are the 10 most variable genes in sci-ATAC-seq dataset\ngenes=np.array(('MSC','RPL31','SERPINB7','ARHGEF4','DCAF8','RNF149','ACMSD','ESPNL','FAM135A','RGS13'))\n\n# Plot the activity of the 10 most variable genes in sci-ATAC-seq dataset\nidxS=np.zeros(genes.shape[0])\nfor ii in range(genes.shape[0]):\n    \n    idx=[geneNames==genes[ii]]\n    idxN=np.squeeze((np.array(idx)))\n    idxS[ii]=np.squeeze(np.asarray(np.where(idxN==True)))\n\nidxSs=idxS.astype(int)\nactData=activityMatrix[:,idxSs]\n\n# Rescale the activity in each cell type from 0 to 1 for plotting\nactDataRe=actData\nfor ik in range(6):\n    actDataRe[ik,:]=rescale(actData[ik,:])\n\nplt.figure(figsize=(12,4))\nax=plt.stem(genes,actDataRe[0,:]) # to see the gene activity in other cells, please change this \n# index to a value from 1 to 5. Default is 0 (B-cell). 1 is for T-cell, 2 for natural killer cells, \n# 3 for Monocytes, 4 for Pneumoctyes, and 5 for Hematopoitic cells.\nplt.xlabel(\"Genes\")\nplt.ylabel(\"Activity score\")\nplt.title('Mean activity of genes in B-cells')\n"},{"cell_type":"markdown","id":"6a97e22d","metadata":{},"source":"##  Multi-omic integration"},{"cell_type":"code","execution_count":null,"id":"36bc8ebd","metadata":{},"outputs":[],"source":"# The pancreatic scRNA-seq data from 5 different technologies (Baron et al, \n# Muraro et al., Xin et al., Wang et al., and Segerstolpe et al.) are first integrated with Seurat. \n# In Seurat, the parameter 'nfeatures' (denoting  the number of features) is set\n# to 2000. The resulting Seurat output is loaded here\ndata = sio.loadmat('../data/outSeurat_pancORG.mat')\noutSeurat=data['outSeurat']\n\n# We now create a genomap for each cell in the data\n# We select the nearest square number to 2000, which is 1936\n# Thus the size of the genomap would be 44 by 44.\n# Next, let us select data with 1936 most variable features \nnumRow=44\nnumCol=44\nvarX=np.var(outSeurat,axis=0)\nidxV=np.argsort(varX)\nidxVX=np.flip(idxV)\noutSeurat1936=outSeurat[:,idxVX[0:numRow*numCol]]\n\n# Construction of the genomaps for the pancreatic data from the five different technologies\ngenoMaps=construct_genomap(outSeurat1936,numRow,numCol,epsilon=0.0,num_iter=200)\n\n\n# Visualize a genomap (we show here the first one (i=0))\nfindI=genoMaps[0,:,:,:]\nplt.figure(4)\nplt.imshow(findI, origin = 'lower',  extent = [0, 10, 0, 10], aspect = 1)\nplt.title('Genomap of a cell from pancreatic dataset')\n# Load the data labels\ngt_data = sio.loadmat('../data/GT_panc.mat')\nGT = np.squeeze(gt_data['GT'])\n\n# Load the index of the training and testing data \n# Here training data are from Baron et al, Muraro et al., Xin et al., and Wang et al.\n#  and the testing data is from Segerstolpe et al.\nindex_data = sio.loadmat('../data/index_panc.mat')\nindxTest = np.squeeze(index_data['indxTest'])\nindxTrain = np.squeeze(index_data['indxTrain'])\nGT = GT - 1 \n\n# Prepare the data for genoNet training\ndataMat_CNNtrain = genoMaps[indxTrain-1,:,:,:] # transfer from matlab indexing to python indexing\ndataMat_CNNtest = genoMaps[indxTest-1,:,:,:]\ngroundTruthTest = GT[indxTest-1]\ngroundTruthTrain = GT[indxTrain-1]\n\nXTrain = dataMat_CNNtrain[:,:,:,:].transpose([0,3,1,2])\nXTest = dataMat_CNNtest.transpose([0,3,1,2])\n# Make labels begin with 0\nyTrain = groundTruthTrain-1\nyTest = groundTruthTest-1\n\nminiBatchSize = 128\n# Train the genoNet\ntorch.manual_seed(0)\nnet = traingenoNet(XTrain, yTrain, maxEPOCH=150, batchSize=miniBatchSize, verbose=True)\n\n# Test the genoNet\ntestset = geneDataset(XTest, yTest)\ntestloader = DataLoader(testset, batch_size=miniBatchSize, shuffle=False)\ndevice = torch.device('cpu')\nprob_test = predict(net, testloader, device)\npred_test = np.argmax(prob_test, axis=-1)\n\nprint('Label transfer accuracy of genomap+genoNet:'+ str(np.sum(pred_test==yTest) / pred_test.shape[0]))"},{"cell_type":"markdown","id":"a0672e72","metadata":{},"source":"## Genomap enables high quality cellular trajectory mapping"},{"cell_type":"code","execution_count":null,"id":"f8d887d6","metadata":{},"outputs":[],"source":"# As cellular trajectory mapping is an unsupervised task, we import the unsupervised genoNet. \n# The difference between a supervised and unsupervised\n# genoNet is the size of the final fully connected (FFC) layer. As unsupervised genoNet\n# is used for feature extraction, its FFC layer needs to have much more neurons than the\n# supervised genoNet. We set the neuron number in the FFC layer of supervised\n# genoNet to 100 and of unsupervised genoNet to 512. \n\n# import the unsupervised genoNet\nfrom genoNetus import genoNet,geneDataset,load_genoNet,predict,traingenoNet \n\n# For the purpose of rapid demonstration, we have created/saved the genomaps for this dataset  \n# The saved genomaps are uploaded here. \n# However, one can also create the genomaps from data/data_proto.mat by\n# following the process described in the third section.\n\n# Load the genomaps and tabular data of proto-vertebrate dataset\ndata = sio.loadmat('../data/genoMaps_Proto.mat')\ngenoMaps = data['genoMaps']\ndataX = sio.loadmat('../data/data_proto.mat')\ndataVec = dataX['X']\n\n# Load data labels for visualization\ngt_data = sio.loadmat('../data/GT_proto.mat')\nGrTruth = np.squeeze(gt_data['GT'])\n\n# Use K-means++ for the initial clustering\namount_initial_centers = 10\ninitial_centers = kmeans_plusplus_initializer(dataVec, amount_initial_centers).initialize()\n# Use x-means for clustering the data\nxmeans_instance = xmeans(dataVec, initial_centers, 20)\nxmeans_instance.process()\n# Extract clustering results: clusters and their centers\nclusters = xmeans_instance.get_clusters()\ncenters = xmeans_instance.get_centers()\n\nclusIndex=np.zeros(dataVec.shape[0]);\nfor i in range(0,len(clusters)):\n    a=(clusters[i])\n    clusIndex[a]=i;\n    \n\n# Set up the training and testing data \ndataMat_CNNtrain = genoMaps # For unsupervised genoNet, all the data are used for training\ndataMat_CNNtest = genoMaps\n\n# Use the estimated labels by x-means clustering for training the unsupervised genoNet\ngroundTruthTest = clusIndex\ngroundTruthTrain = clusIndex\nclassNum = len(np.unique(groundTruthTrain))\nXTrain = dataMat_CNNtrain[:,:,:,:].transpose([3,2,0,1])\nXTest = dataMat_CNNtest.transpose([3,2,0,1])\nyTrain = groundTruthTrain\nyTest = groundTruthTest\n\n# Train the genoNet\nminiBatchSize = 128\n\n# For the purpose of rapid demonstration, we have trained the genoNet model, which\n# is uploaded here. But if one wants to train the \n# genoNet on his/her dataset, please uncomment the following code:\n# net = traingenoNet(XTrain, yTrain, maxEPOCH=30, batchSize=miniBatchSize, verbose=True)\n\n# Load the trained genoNet. \nnet = load_genoNet([1,27,27], 20, '../data/genoNet_PHATE_ZF.pt')\n\n# Extract genoNet features from the final fully connected layer\n# for PHATE analysis. \ngx=np.reshape(XTrain,(XTrain.shape[0],1,XTrain.shape[2],XTrain.shape[3]))\ndevice = _get_device()\nt = torch.from_numpy(gx).to(device)\nnet=net.double()\ndataAtFC=net.forwardX(t)\ndata=dataAtFC.cpu().detach().numpy()\n\n# Run PHATE on the genoNet features\nphate_op = phate.PHATE(random_state=1)\nX_embedded = phate_op.fit_transform(data)\n\n# Plot embeddings\nplt.figure(11)\nscatter=plt.scatter(X_embedded[:,0], X_embedded[:,1],s=2,c=GrTruth,cmap=\"jet\")\nplt.xlabel(\"PHATE1\")\nplt.ylabel(\"PHATE2\")\nplt.title('PHATE embedding of genoNet features')\nlegend1 = plt.legend(*scatter.legend_elements(),\n                    loc=\"lower left\", title=\"Time points\")\nplt.show()\n\n# Run PHATE on the raw data and compare with result from the proposed approach\nX_embedded = phate_op.fit_transform(dataVec)\n\n# Plot embeddings\nplt.figure(12)\nscatter=plt.scatter(X_embedded[:,0], X_embedded[:,1],s=2,c=GrTruth,cmap=\"jet\")\nplt.xlabel(\"PHATE1\")\nplt.ylabel(\"PHATE2\")\nplt.title('PHATE embedding of raw data')\nlegend1 = plt.legend(*scatter.legend_elements(),\n                    loc=\"lower left\", title=\"Time points\")\nplt.show()"},{"cell_type":"markdown","id":"737148e3","metadata":{},"source":"## Dimensionality reduction, visualization and clustering"},{"cell_type":"code","execution_count":null,"id":"8a8aab70","metadata":{},"outputs":[],"source":"# We use unsupervised genoNet for dimensionality reduction.\n# We create the genomaps of a dataset and then train the unsupervised genoNet\n# using the genomaps. We then extract features from the final fully connected layer of the genoNet, which\n# have a lower dimensionality than the original data. We then apply\n# t-SNE on the features to obtain the embeddings. \n\n# Load the genomaps and tabular data of comprehensive classification of mouse retinal bipolar cells \ndata = sio.loadmat('../data/genoMap_comClass.mat')\ngenoMaps = data['genoMaps']\ndataX = sio.loadmat('../data/data_comClass.mat')\ndataVec = dataX['X']\n\n# Load cell labels for visualization\ngt_data = sio.loadmat('../data/GT_comClass.mat')\nGrTruth = np.squeeze(gt_data['GT'])\n\n# Use k-means++ for initial clustering\namount_initial_centers = 10\ninitial_centers = kmeans_plusplus_initializer(dataVec, amount_initial_centers).initialize()\n# Use x-means for clustering the data\nxmeans_instance = xmeans(dataVec, initial_centers, 20)\nxmeans_instance.process()\n# Extract clustering results: clusters and their centers\nclusters = xmeans_instance.get_clusters()\ncenters = xmeans_instance.get_centers()\n\nclusIndex=np.zeros(dataVec.shape[0]);\nfor i in range(0,len(clusters)):\n    a=(clusters[i])\n    clusIndex[a]=i;\n    \n\n# Set up the training and testing data \ndataMat_CNNtrain = genoMaps # For unsupervised genoNet, all the data are used for training\ndataMat_CNNtest = genoMaps\n\n# Use the estimated labels for genoNet training\ngroundTruthTest = clusIndex\ngroundTruthTrain = clusIndex\nclassNum = len(np.unique(groundTruthTrain))\n\n# Prepare data for training \nXTrain = dataMat_CNNtrain[:,:,:,:].transpose([3,2,0,1])\nXTest = dataMat_CNNtest.transpose([3,2,0,1])\nyTrain = groundTruthTrain\nyTest = groundTruthTest\n\n# Train the genoNet\nminiBatchSize = 128\n\n# If one wants to train the genoNet on his/her dataset, please uncomment the folloiwng code line\n# net = traingenoNet(XTrain, yTrain, maxEPOCH=30, batchSize=miniBatchSize, verbose=True)\n\n# Load the trained genoNet\nnet = load_genoNet([1,33,33], 19, '../data/genoNet_TSNE_ComClass.pt')\n# Extract genoNet features from the final fully connected layer\ngx=np.reshape(XTrain,(XTrain.shape[0],1,XTrain.shape[2],XTrain.shape[3]))\ndevice = _get_device()\nt = torch.from_numpy(gx).to(device)\nnet=net.double()\ndataAtFC=net.forwardX(t)\n\n# Run t-SNE on the genoNet features\nX=dataAtFC.cpu().detach().numpy()\nX_embedded = TSNE(n_components=2, learning_rate='auto',init='random').fit_transform(X)\n\nplt.figure(13)\nscatter=plt.scatter(X_embedded[:,0], X_embedded[:,1],s=2,c=GrTruth,cmap=\"jet\")\nplt.xlabel(\"tSNE1\")\nplt.ylabel(\"tSNE2\")\nplt.title('t-SNE embedding of genoNet features')\nlegend1 = plt.legend(*scatter.legend_elements(),\n                    loc=\"lower left\", title=\"Classes\")\nplt.show()\n\n\nX_embedded = TSNE(n_components=2, learning_rate='auto',init='random').fit_transform(dataVec)\n\nplt.figure(14)\nscatter=plt.scatter(X_embedded[:,0], X_embedded[:,1],s=2,c=GrTruth,cmap=\"jet\")\nplt.xlabel(\"tSNE1\")\nplt.ylabel(\"tSNE2\")\nplt.title('t-SNE embedding of raw data')\nlegend1 = plt.legend(*scatter.legend_elements(),\n                    loc=\"lower left\", title=\"Classes\")\nplt.show()\n\n"}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.8"}},"nbformat":4,"nbformat_minor":5}